tests/unit/test_id_cli.py

```diff
@@ -0,0 +1,528 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
+from pathlib import Path
+import pytest
+
+from cylc.flow.async_util import pipe
+from cylc.flow.exceptions import UserInputError, WorkflowFilesError
+from cylc.flow.id import detokenise, tokenise, Tokens
+from cylc.flow.id_cli import (
+    _expand_workflow_tokens,
+    _parse_src_path,
+    _validate_constraint,
+    _validate_workflow_ids,
+    _validate_number,
+    parse_ids_async,
+)
+from cylc.flow.pathutil import get_cylc_run_dir
+from cylc.flow.workflow_files import WorkflowFiles
+
+
+@pytest.fixture(scope='module')
+def abc_src_dir(tmp_path_factory):
+    """Src dir containing three workflows, a, b & c."""
+    cwd_before = Path.cwd()
+    tmp_path = tmp_path_factory.getbasetemp()
+    os.chdir(tmp_path)
+    for name in ('a', 'b', 'c'):
+        Path(tmp_path, name).mkdir()
+        Path(tmp_path, name, WorkflowFiles.FLOW_FILE).touch()
+    yield tmp_path
+    os.chdir(cwd_before)
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    'ids_in,ids_out',
+    [
+        (('a//',), ['a']),
+        (('a//', 'a//'), ['a']),
+        (('a//', 'b//'), ['a', 'b']),
+    ]
+)
+async def test_parse_ids_workflows(ids_in, ids_out):
+    """It should parse workflows & tasks."""
+    workflows, _ = await parse_ids_async(*ids_in, constraint='workflows')
+    assert list(workflows) == ids_out
+    assert list(workflows.values()) == [[] for _ in workflows]
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    'ids_in,ids_out',
+    [
+        (('./a',), ['a']),
+    ]
+)
+async def test_parse_ids_workflows_src(ids_in, ids_out, abc_src_dir):
+    """It should parse src workflows."""
+    workflows, _ = await parse_ids_async(
+        *ids_in,
+        src=True,
+        constraint='workflows',
+    )
+    assert list(workflows) == ids_out
+    assert list(workflows.values()) == [[] for _ in workflows]
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    'ids_in,ids_out',
+    [
+        (
+            ('a//i',),
+            {'a': ['//i']},
+        ),
+        (
+            ('a//i', 'a//j'),
+            {'a': ['//i', '//j']},
+        ),
+        (
+            ('a//i', 'b//i'),
+            {'a': ['//i'], 'b': ['//i']},
+        ),
+        (
+            ('a//', '//i', 'b//', '//i'),
+            {'a': ['//i'], 'b': ['//i']},
+        ),
+    ]
+)
+async def test_parse_ids_tasks(ids_in, ids_out):
+    """It should parse workflow tasks in two formats."""
+    workflows, _ = await parse_ids_async(*ids_in, constraint='tasks')
+    assert {
+        workflow_id: [detokenise(tokens) for tokens in tokens_list]
+        for workflow_id, tokens_list in workflows.items()
+    } == ids_out
+
+
+@pytest.mark.parametrize(
```

Comment:
1 slipped through the net
```suggestion
@pytest.mark.asyncio
@pytest.mark.parametrize(
```